homeassistant/util/executor.py

```diff
@@ -0,0 +1,108 @@
+from threading import Thread
+import time
+import traceback
+
+from homeassistant.util.thread import async_raise
+
+_LOGGER = logging.getLogger(__name__)
+
+MAX_LOG_ATTEMPTS = 2
+
+_JOIN_ATTEMPTS = 10
+
+EXECUTOR_SHUTDOWN_TIMEOUT = 10
+
+
+def _log_thread_running_at_shutdown(name: str, ident: int) -> None:
+    """Log the stack of a thread that was still running at shutdown."""
+    frames = sys._current_frames()  # pylint: disable=protected-access
+    stack = frames.get(ident)
+    formatted_stack = traceback.format_stack(stack)
+    _LOGGER.warning(
+        "Thread[%s] is still running at shutdown: %s",
+        name,
+        "".join(formatted_stack).strip(),
+    )
+
+
+def join_or_interrupt_threads(
+    threads: set[Thread], timeout: float, log: bool
+) -> set[Thread]:
+    """Attempt to join or interrupt a set of threads."""
+    joined = set()
+    timeout_per_thread = timeout / len(threads)
+
+    for thread in threads:
+        thread.join(timeout=timeout_per_thread)
+
+        if not thread.is_alive() or thread.ident is None:
+            joined.add(thread)
+            continue
+
+        if log:
+            _log_thread_running_at_shutdown(thread.name, thread.ident)
+
+        async_raise(thread.ident, SystemExit)
+
+    return joined
+
+
+class InterruptibleThreadPoolExecutor(ThreadPoolExecutor):
+    """A ThreadPoolExecutor instance that will not deadlock on shutdown."""
+
+    def logged_shutdown(self) -> None:
+        """Shutdown backport from cpython 3.9 with interrupt support added."""
+        with self._shutdown_lock:  # type: ignore[attr-defined]
+            self._shutdown = True
+            # Drain all work items from the queue, and then cancel their
+            # associated futures.
+            while True:
+                try:
+                    work_item = self._work_queue.get_nowait()
+                except queue.Empty:
+                    break
+                if work_item is not None:
+                    work_item.future.cancel()
+            # Send a wake-up to prevent threads calling
+            # _work_queue.get(block=True) from permanently blocking.
+            self._work_queue.put(None)
+
+        # The above code is backported from python 3.9
+        #
+        # For maintainability join_threads_or_timeout is
+        # a separate function since it is not a backport from
+        # cpython itself
+        #
+        self.join_threads_or_timeout()
+
+    def join_threads_or_timeout(self) -> None:
+        """Join threads or timeout."""
+        remaining_threads = set(self._threads)  # type: ignore[attr-defined]
+        start_time = time.monotonic()
+        timeout_remaining: float = EXECUTOR_SHUTDOWN_TIMEOUT
+        attempt = 0
+
+        while True:
+            if not remaining_threads:
+                return
+
+            attempt += 1
+
+            remaining_threads -= join_or_interrupt_threads(
+                remaining_threads,
+                timeout_remaining / _JOIN_ATTEMPTS,
```

Comment:
Since we're returning after the time remaining goes to zero or no threads remaining, it means that we could be attempting to join more times than the join attempt constant. I don't think that's a problem. It is just an observation. Maybe we shouldrremain the constant to `_MINIMUM_JOIN_ATTEMPTS`?

Reply:
Although it won't attempt all attempts if all threads finish early, so maybe keep the name anyway?

Reply:
> Although it won't attempt all attempts if all threads finish early, so maybe keep the name anyway?

Naming to show intent is hard on this one since it doesn't seem like there is a good description that can be boiled down into something short. Probably best to keep as is unless there is an aha naming movement.